FuelLabs/fuel-indexer
# enhancement: add sorting and offset-based pagination to GraphQL API queries

Closes #656.

## Changelog
- Add doc comments in a number of places
- Rename a number of structs and variables for clarity
- Add `QueryParams` struct to associate potential parameters together
- Add `to_sql()` method for each query parameter type
- Add support for `order` keyword; this allows for sorting a list of results by a field
- Add support for `first` keyword; this allows for limiting a result set to the first N records
- Add support for `offset` keyword; this allows for skipping the first N records in a result set
- Add tests for new keywords

## Testing Plan
CI should pass.

### Manual Testing
0. Compile fresh WASM modules and clear database.
1. `cargo run --bin fuel-indexer -- run --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --manifest examples/block-explorer/explorer-indexer/explorer_indexer.manifest.yaml`
2. Open the corresponding playground: `http://localhost:29987/api/playground/fuel_examples/explorer_indexer`
3. Run the following query. You should see that the last 10 blocks are shown in reverse order of block height.
```
query {
  block(order: {desc: height}, first: 10) {
    id
    height
    timestamp
    hash
  }
}
```
4. Run the following query. You should see that the next 5 blocks after the first 50 are shown in chronological order of block height.
```
query {
  block(order: {asc: height}, first: 5, offset: 50) {
    id
    height
    timestamp
    hash
  }
}
```
